jsonpatcher/jsonpatcher.go

```diff
@@ -91,66 +80,37 @@ func (jp *jsonPatcher) Create(actions []core.Action) ([]core.Event, format.Node,
 	return events, n, nil
 }
 
-func (jp *jsonPatcher) Reduce(events []core.Event, datastore ds.TxnDatastore, baseKey ds.Key) ([]core.ReduceAction, error) {
-	txn, err := datastore.NewTransaction(false)
-	if err != nil {
-		return nil, err
+func (jp *jsonPatcher) Reduce(e core.Event, oldState []byte) (*core.CodecResult, error) {
+	je, ok := e.(patchEvent)
+	if !ok {
+		return nil, fmt.Errorf("event unrecognized for jsonpatcher eventcodec")
 	}
-	defer txn.Discard()
-
-	actions := make([]core.ReduceAction, len(events))
-	for i, e := range events {
-		je, ok := e.(patchEvent)
-		if !ok {
-			return nil, fmt.Errorf("event unrecognized for jsonpatcher eventcodec")
-		}
-		key := baseKey.ChildString(e.Model()).ChildString(e.EntityID().String())
-		switch je.Patch.Type {
-		case create:
-			exist, err := txn.Has(key)
-			if err != nil {
-				return nil, err
-			}
-			if exist {
-				return nil, errCantCreateExistingInstance
-			}
-			if err := txn.Put(key, je.Patch.JSONPatch); err != nil {
-				return nil, fmt.Errorf("error when reducing create event: %v", err)
-			}
-			actions[i] = core.ReduceAction{Type: core.Create, Model: e.Model(), EntityID: e.EntityID()}
-			log.Debug("\tcreate operation applied")
-		case save:
-			value, err := txn.Get(key)
-			if errors.Is(err, ds.ErrNotFound) {
-				return nil, errSavingNonExistentInstance
-			}
-			if err != nil {
-				return nil, err
-			}
-			patchedValue, err := jsonpatch.MergePatch(value, je.Patch.JSONPatch)
-			if err != nil {
-				return nil, fmt.Errorf("error when reducing save event: %v", err)
-			}
-			if err = txn.Put(key, patchedValue); err != nil {
-				return nil, err
-			}
-			actions[i] = core.ReduceAction{Type: core.Save, Model: e.Model(), EntityID: e.EntityID()}
-			log.Debug("\tsave operation applied")
-		case delete:
-			if err := txn.Delete(key); err != nil {
-				return nil, err
-			}
-			actions[i] = core.ReduceAction{Type: core.Delete, Model: e.Model(), EntityID: e.EntityID()}
-			log.Debug("\tdelete operation applied")
-		default:
-			return nil, errUnknownOperation
+	switch je.Patch.Type {
```

Comment:
So now, we essentially don't mutate _any_ state within the codec. This is now just a pure function that takes the initial state, and returned an updated state. This is slightly more akin to something like Redux in that sense...
I think this makes it easier for external devs to implement their own EventCodecs...